### ISSUE TITLE
Analysis extension API

### ISSUE THREAD

**Issue by mansellan** [WaynePhillipsEA/twinbasic#51]
**Is your feature request related to a problem?**
Third party code analysis projects could provide significant benefit to twinBASIC if an API were available to access the AST.

**Describe the solution you'd like**
Projects like [Rubberduck](https://github.com/rubberduck-vba) (RD) can perform in-depth analysis over code, highlighting issues and suggesting automated fixes. In the case of RD, this entails performing all lexing, parsing and symbol resolution from the actual source code, as the compiler internal data structures are not directly available. This is a wasteful duplication of effort already performed by the compiler.
If twinBASIC could provide an API which allowed access to the abstract syntax tree, analysis could be performed directly by third-party extensions without such a performance overhead.

**Additional context**
I have no idea whether this is feasible without exposing excessive information of the compiler internals.

**Comment by retailcoder** [comment]
I thought of proposing this, but didn't because a lot of what Rubberduck's static code analysis does is really what the VBA/VB6 compiler's job should be, and I think it's important and necessary that the twinBASIC compiler _itself_ implements as much diagnostics as possible.

I wouldn't want to put a hypothetical "Rubberduck Twin" tB extension into a situation where its hard-worked diagnostics and features get integrated into the tB compiler one by one, until the extension is just irrelevant.

On the other hand, twinBASIC could likely benefit in being spliced into multiple separate projects, with the core compiler services behind closed doors but the diagnostics and refactoring extensions open-sourced: extracting the static code analysis from the compiler itself would make excellent dogfood for such an extensibility API.

**Comment by WaynePhillipsEA** [comment]
I do really like the thought of exposing this API.   What I'm struggling with is the cost to implement (in terms of time) vs. the cost to just implement the inspections within the compiler directly.  It might be worth trying to come up with a list of say the top 5 wish-list  for inspections, and from that I can probably get a better idea of the time cost for the in-built compiler option.

It would be really cool to have the inspection extensions themselves implemented in a tB project...

**Comment by bclothier** [comment]
Agreed. I think Roslyn is the model you want to emulate for tB. That would in turn allow people to write their own analyzer and thus enforce project-specific requirement in addition to having inspections that is broadly applicable.

**Comment by IvenBach** [comment]
- 0 Option Explicit - No exceptions allowed!
- 1 Default member implicitly used
- 2 Implicit type `As Variant` when no type is specified in declaration
- 3 Implicit ByRef when no explicit pass-by is given
- 4 Default CodeName `Sheet1` rename opportunity
- 5 Unused identifiers

The above items that RD brought to my attention when I started using it made me think about code critically. So many “helpful” implicit things that make it harder to understand what was going on. Very hard to learn about this stuff if you don’t know is happening.

**Comment by WaynePhillipsEA** [comment]
@IvenBach now that's a proper programmer, staying his list count at zero!   All those on your list (ignoring the Excel be specific one) could be implemented very, very easily in the compiler and that suggests to me that at this moment in time we would be better off just implementing the inspections inside the compiler for now.  Long term, we can certainly look at exposing this API because I think it will be amazing once available.

So for now I think we should start opening individual feature requests for the inspections that you would like to see implemented.
